src/Support/InferExtensions/ResponderTypeInfer.php

```diff
@@ -0,0 +1,128 @@
+use Illuminate\Support\Str;
+
+class ResponderTypeInfer implements ExpressionTypeInferExtension
+{
+    private $modelType;
+    private $modelPropertyName;
+    public static $transformerModelTypesCache = [];
+
+    public function getType(Expr $node, Scope $scope): ?Type
+    {
+        if (
+            !$scope->classDefinition()?->isInstanceOf(TransformerAbstract::class)
+            || $scope->classDefinition()?->name === TransformerAbstract::class
+        ) {
+            return null;
+        }
+
+        $transformerArguments = $scope->classDefinition()->getMethodDefinition('transform')->type->arguments;
+
+        $this->modelPropertyName = array_key_first($transformerArguments);
+        $this->modelType = $transformerArguments[$this->modelPropertyName]->is;
+
+        /** $this->? */
+        if (
+            $node instanceof Node\Expr\PropertyFetch && ($node->var->name ?? null) === $this->modelPropertyName
+            && is_string($node->name->name ?? null)
+            && !array_key_exists($node->name->name, $scope->classDefinition()->properties)
+            && ($type = $this->modelType($scope->classDefinition(), $scope))
+        ) {
+            return $scope->getPropertyFetchType($type, $node->name->name);
+        }
+
+        /**
+         * new MissingValue
+         */
+        if ($scope->getType($node)->isInstanceOf(MissingValue::class)) {
+            return new ObjectType(MissingValue::class);
+        }
+
+        return null;
+    }
+
+    private function modelType(ClassDefinition $jsonClass, Scope $scope): ?Type
+    {
+        if ([$cachedModelType, $cachedModelDefinition] = static::$transformerModelTypesCache[$jsonClass->name] ?? null) {
+            if ($cachedModelDefinition) {
+                $scope->index->registerClassDefinition($cachedModelDefinition);
+            }
+
+            return $cachedModelType;
+        }
+
+        $modelClass = $this->getModelName(
+            $jsonClass->name,
+            new \ReflectionClass($jsonClass->name),
+            $scope->nameResolver,
+        );
+
+        $modelType = new UnknownType("Cannot resolve [$modelClass] model type.");
+        $modelClassDefinition = null;
+        if ($modelClass && is_a($modelClass, Model::class, true)) {
+            try {
+                $modelClassDefinition = (new ModelInfo($modelClass))->type();
+
+                $scope->index->registerClassDefinition($modelClassDefinition);
+
+                $modelType = new ObjectType($modelClassDefinition->name);
+            } catch (\LogicException $e) {
+                // Here doctrine/dbal is not installed.
+                $modelType = null;
+                $modelClassDefinition = null;
+            }
+        }
+
+        static::$transformerModelTypesCache[$jsonClass->name] = [$modelType, $modelClassDefinition];
+
+        return $modelType;
+    }
+
+    private function getModelName(string $jsonResourceClassName, \ReflectionClass $reflectionClass, FileNameResolver $getFqName)
```

Comment:
What does `getFqName` mean? Can we change it to something more meaningful?

Reply:
getFullyQuallifiedName, library standard

Reply:
Ah, gotcha!